go/sig3/sig3.go

```diff
@@ -0,0 +1,401 @@
+package sig3
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"github.com/keybase/client/go/kbcrypto"
+	"github.com/keybase/client/go/msgpack"
+	"github.com/keybase/go-crypto/ed25519"
+)
+
+//
+
+// Generic sig3 wrapper class, should implement the following interface.
+type Generic interface {
+	Signer() *Signer
+	Prev() *LinkID
+	Seqno() Seqno
+	Outer() OuterLink
+	Inner() *InnerLink
+
+	// Inner methods for generic decoding of incoming sig3 links
+	outerPointer() *OuterLink
+	setVerifiedBit()
+	verify() error
+}
+
+// Base struct for sig3 links that contains much of the raw material pulled down or off local storage.
+// Most implementations of sig3 links should include this base class.
+type Base struct {
+	verified bool
```

Comment:
Does this mean just that the signature is verified against the given kid (which is in Base.verify) or also link-specific things like reverse sigs were verified? e.g., would it be set in RotateKey.verify() if the base sig checks out (because the bit is set at the end of Base.verify()) but the reverse sigs don't?

Reply:
It's all or nothing. It should only be unset in the case of a stubbed link (since we can't verify) or if there was an error in any crypto operations, including outer and inner sigs.